### DISCUSSION TITLE
VertiPaq Analyzer Enhancement for Missing Keys and Invalid Rows

### DISCUSSION THREAD

**Discussion by corycundy** [TabularEditor/TabularEditor3#927] [Ideas]
A small enhancement to the VertiPaq Analyzer UI could provide quick insight into Missing Keys and Invalid Rows. When expanding a Table in the relationship tab and right clicking on a row (relationship) with a non zero value in the Missing Keys or Invalid Rows column, the context menu could provide two more options. One would be to open a new DAX query tab with a query to show Missing Keys and the other for Invalid Rows. The queries might look something like the examples below, but with the tables and columns being replaced from the information on the relationship that was right clicked on. The SAMPLE function could be integrated with how other queries SAMPLE the first 1000 rows with the ability to run the full query and be removed from these queries.

I believe this would provide a lot of value as I have found that users like that VertiPaq Analyzer can easily identifying that there are Missing Keys and Invalid Rows, but have a harder time finding what they all are other than using the Sample Violations. The Sample Violations is generally meant to only show a smaller number even though configurable. I have suggested this feature to Darren many years ago, but it did not make it into DAX Studio.

//Missing Keys Query
EVALUATE
CALCULATETABLE(
    SAMPLE(1000, DISTINCT('Fact Table'[DimTableKey]), 'Fact Table'[DimTableKey], ASC),
    ISBLANK('DimTable'[DimTableKey])
)

//Invalid Rows Query
EVALUATE
CALCULATETABLE(
    SAMPLE(1000, 'Fact Table', 'Fact Table'[DimTableKey], ASC),
    ISBLANK('DimTable'[DimTableKey])
)

**Comment by otykier** [comment]
Thanks for the suggestion.

While we can certainly add such functionality to TE3, I wonder if it would make sense to share some of the logic (i.e. to generate the DAX query) within the VertiPaq Analyzer libs? What do you think, @marcosqlbi @albertospelta

**Comment by marcosqlbi** [comment]
We evaluated the effort involved in exposing the feature from VPAX libraries but it doesn't worth the effort.
@otykier you can use the same DAX code we generate here to get the extraction for any relationship (also inactive one) and any compatibility level (different code for SSAS 2012/2014): https://github.com/sql-bi/VertiPaq-Analyzer/blob/54713f4de160fe5351dca9b83803f354cb003917/src/Dax.Model.Extractor/StatExtractor.cs#L106
